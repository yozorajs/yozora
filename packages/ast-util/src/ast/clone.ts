import type { IYastNode, IYastParent, Root } from '@yozora/ast'

/**
 * Shallow clone ast until the match reaches the termination condition.
 * Note that the root node will not be traversed, that is, the root node will
 * never be passed into the `endCondition` function.
 *
 * @param root
 * @param endCondition
 * @returns
 */
export function shallowCloneAst(
  root: Root,
  endCondition: (node: IYastNode, parent: IYastParent, childIndex: number) => boolean,
): Root {
  const clone = (u: IYastParent): IYastParent => {
    const nextChildren = []
    const { children } = u
    for (let i = 0; i < children.length; ++i) {
      const v = children[i] as IYastParent
      if (endCondition(v, u, i)) break

      const nextChild = v.children == null ? v : clone(v)
      nextChildren.push(nextChild)
    }
    return { ...u, children: nextChildren }
  }
  return clone(root) as Root
}
