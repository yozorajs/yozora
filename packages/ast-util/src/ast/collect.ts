import type { Root, YastNode, YastNodeType, YastParent } from '@yozora/ast'
import { createNodeTypeMatcher } from './util'

/**
 * Collect nodes of the specified type through pre-order traversal.
 *
 * Note that if the matched node has other matching descendant nodes, these
 * descendant nodes will not be collected repeatedly, that is, only the
 * currently matched node will be collected, and will not enter the subtree for
 * collection afterwards.
 *
 * @param root
 * @param aimTypes
 */
export function collectNodes<T extends YastNodeType, O extends YastNode<T>>(
  root: Root,
  aimTypes: ReadonlyArray<T>,
): O[] {
  const nodes: O[] = []
  const isMatch = createNodeTypeMatcher(aimTypes)
  const collect = (u: YastParent): void => {
    const { children } = u
    if (children != null) {
      for (const v of children) collect(v as YastParent)
    }
    if (isMatch(u)) nodes.push(u as unknown as O)
  }
  collect(root)
  return nodes
}
