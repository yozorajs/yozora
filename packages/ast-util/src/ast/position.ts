import type { Node, Parent, Root } from '@yozora/ast'

export function removePositions(immutableAst: Readonly<Root>): Root {
  function remove(node: Node): Node {
    const { position, children, ...nextNode } = node as Parent
    for (const key of Object.keys(nextNode)) {
      nextNode[key] = Array.isArray(nextNode[key]) ? nextNode[key].map(remove) : nextNode[key]
    }
    if (children) {
      ;(nextNode as Parent).children = children.map(remove)
    }
    return nextNode
  }
  const tidyAst = remove(immutableAst) as Root
  return tidyAst
}
