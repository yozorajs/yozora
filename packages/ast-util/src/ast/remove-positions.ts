import type { IRoot, IYastNode, IYastParent } from '@yozora/ast'

export function removePositions(immutableAst: Readonly<IRoot>): IRoot {
  function remove(node: IYastNode): IYastNode {
    const { position, children, ...nextNode } = node as IYastParent
    if (children) {
      ;(nextNode as IYastParent).children = children.map(remove)
    }
    return nextNode
  }
  const tidyAst = remove(immutableAst) as IRoot
  return tidyAst
}
