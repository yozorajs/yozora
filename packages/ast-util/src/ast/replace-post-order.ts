import type { Node, NodeType, Parent, Root } from '@yozora/ast'
import type { INodeMatcher } from './util'
import { createNodeMatcher, createShallowNodeCollector } from './util'

/**
 * Traverse AST and replace nodes in post-order.
 *
 * Note that the root node will not be traversed, that is, the root node will
 * never be passed into the `replace` function as the first parameter.
 *
 * 需要注意的是，尽管此函数不会直接操作原AST，但它会尽可能地复用未改变的节点；
 * 此外它依赖于 replace 函数将传入的节点视作不可修改的，所以在使用时，需要注意不
 * 要在 replace 中直接修改节点，如果需要修改，应该返回一个新的节点（节点属性可以
 * 复用）
 *
 * It should be noted that although this function does not directly manipulate
 * the original AST, it will reuse unchanged nodes as much as possible. In
 * additional, it relies on the passed-in `replace` function to treat the passed
 * node as immutable, so when using this function, you need to be careful not
 * to modify the node directly in `replace` callback. If you need to perform
 * some changes, you should return a new node (node attributes can be reused).
 *
 * @param immutableRoot
 * @param aimTypesOrNodeMatcher
 * @param replace
 */
export function shallowMutateAstInPostorder(
  immutableRoot: Readonly<Root>,
  aimTypesOrNodeMatcher: ReadonlyArray<NodeType> | INodeMatcher | null,
  replace: (
    immutableNode: Readonly<Node>,
    immutableParent: Readonly<Parent>,
    childIndex: number,
  ) => Node | Node[] | null,
): Readonly<Root> {
  const isMatched: INodeMatcher = createNodeMatcher(aimTypesOrNodeMatcher)

  const traverse = (children: ReadonlyArray<Node>, parent: Readonly<Parent>): Node => {
    // Recursively processing the descendant nodes in post-order traverse.
    const collector0 = createShallowNodeCollector(children as Node[])
    for (let i = 0; i < children.length; ++i) {
      const child = children[i] as Parent
      const subChildren: ReadonlyArray<Node> = child.children

      // Whether to process the subtree recursively.
      const nextChild = subChildren && subChildren.length > 0 ? traverse(subChildren, child) : child
      collector0.add(nextChild, child, i)
    }

    // Processing current layer of nodes.
    const nextChildren: Node[] = collector0.collect()
    const collector1 = createShallowNodeCollector(nextChildren)
    for (let i = 0; i < nextChildren.length; ++i) {
      const child = nextChildren[i]
      const nextChild = isMatched(child) ? replace(child, parent, i) : child
      collector1.add(nextChild, child, i)
    }

    const finalChildren: Node[] = collector1.collect()
    const result: Node =
      finalChildren === children ? parent : { ...parent, children: finalChildren }
    return result
  }
  return traverse(immutableRoot.children, immutableRoot) as Root
}
