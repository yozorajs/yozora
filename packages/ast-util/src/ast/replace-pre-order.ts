import type { Node, NodeType, Parent, Root } from '@yozora/ast'
import type { INodeMatcher } from './util'
import { createNodeMatcher, createShallowNodeCollector } from './util'

/**
 * Traverse AST and replace nodes in pre-order.
 *
 * Note that the root node will not be traversed, that is, the root node will
 * never be passed into the `replace` function as the first parameter.
 *
 * 需要注意的是，尽管此函数不会直接操作原AST，但它会尽可能地复用未改变的节点；
 * 此外它依赖于 replace 函数将传入的节点视作不可修改的，所以在使用时，需要注意不
 * 要在 replace 中直接修改节点，如果需要修改，应该返回一个新的节点（节点属性可以
 * 复用）
 *
 * It should be noted that although this function does not directly manipulate
 * the original AST, it will reuse unchanged nodes as much as possible. In
 * additional, it relies on the passed-in `replace` function to treat the passed
 * node as immutable, so when using this function, you need to be careful not
 * to modify the node directly in `replace` callback. If you need to perform
 * some changes, you should return a new node (node attributes can be reused).
 *
 * @param immutableRoot
 * @param aimTypesOrNodeMatcher
 * @param replace
 */
export function shallowMutateAstInPreorder(
  immutableRoot: Readonly<Root>,
  aimTypesOrNodeMatcher: ReadonlyArray<NodeType> | INodeMatcher | null,
  replace: (
    immutableNode: Readonly<Node>,
    immutableParent: Readonly<Parent>,
    childIndex: number,
  ) => Node | Node[] | null,
): Readonly<Root> {
  const isMatched: INodeMatcher = createNodeMatcher(aimTypesOrNodeMatcher)

  const traverse = (children: ReadonlyArray<Node>, parent: Readonly<Parent>): Node => {
    // Processing current layer of nodes.
    const collector = createShallowNodeCollector(children as Node[])
    for (let i = 0; i < children.length; ++i) {
      const child = children[i] as Parent
      if (isMatched(child)) {
        const nextChild = replace(child, parent, i)
        collector.add(nextChild, child, i)
      } else {
        // Recursively processing the descendant nodes in pre-order traverse.
        const subChildren: ReadonlyArray<Node> = child.children

        // Whether to process the subtree recursively.
        const nextChild: Node =
          subChildren && subChildren.length > 0 ? traverse(subChildren, child) : child
        collector.add(nextChild, child, i)
      }
    }

    const finalChildren: Node[] = collector.collect()
    const result: Node =
      finalChildren === children ? parent : { ...parent, children: finalChildren }
    return result
  }
  return traverse(immutableRoot.children, immutableRoot) as Root
}
