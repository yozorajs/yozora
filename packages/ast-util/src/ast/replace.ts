import type { Root, YastNode, YastNodeType, YastParent } from '@yozora/ast'

/**
 * Traverse AST and replace nodes.
 * @param root
 * @param aimTypes
 * @param replace
 */
export function replaceAST(
  root: Root,
  aimTypes: ReadonlyArray<YastNodeType> | null,
  replace: (
    node: Readonly<YastNode>,
    parent: Readonly<YastParent>,
    childIndex: number,
  ) => YastNode | YastNode[] | void,
): void {
  const visit = (o: YastNode): void => {
    const u: YastParent = o as YastParent

    // Recursively visit.
    if (u.children != null) {
      const { children } = u
      let nextChildren: YastNode[] | null = null
      for (let i = 0; i < children.length; ++i) {
        const v = children[i]
        let replaced = false
        if (aimTypes == null || aimTypes.indexOf(v.type) > -1) {
          const newChild = replace(v, u, i)
          if (newChild != null) {
            replaced = true
            if (nextChildren == null) nextChildren = children.slice(0, i)
            if (Array.isArray(newChild)) nextChildren.push(...newChild)
            else nextChildren.push(newChild)
          }
        }

        if (!replaced) {
          visit(v)
          if (nextChildren != null) nextChildren.push(v)
        }
      }

      // Replace children if any of them has been replaced.
      if (nextChildren != null) {
        u.children = nextChildren
      }
    }
  }
  visit(root)
}
