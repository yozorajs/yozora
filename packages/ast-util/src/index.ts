export * from './ast/clone'
export * from './ast/collect'
export * from './ast/replace-post-order'
export * from './ast/replace-pre-order'
export * from './ast/search'
export * from './ast/toc'
export * from './ast/traverse'
export * from './ast/util'
export * from './definition'
export * from './excerpt'
export * from './footnote-definition'
export * from './url'
