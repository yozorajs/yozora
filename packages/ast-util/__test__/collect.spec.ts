import type { Root } from '@yozora/ast'
import { ListType, ParagraphType, TextType } from '@yozora/ast'
import { loadJSONFixture } from 'jest.setup'
import { collectNodes } from '../src'

describe('collectNodes', function () {
  test('basic1', function () {
    const originalAst: Readonly<Root> = loadJSONFixture('basic1.ast.json')
    const ast: Root = loadJSONFixture('basic1.ast.json')

    expect(collectNodes(ast, [TextType])).toMatchSnapshot()
    expect(collectNodes(ast, [ParagraphType, ListType])).toMatchSnapshot()
    expect(ast).toEqual(originalAst)
  })
})
