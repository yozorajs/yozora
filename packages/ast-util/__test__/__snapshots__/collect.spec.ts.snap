// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collectNodes basic1 1`] = `
Array [
  Object {
    "type": "text",
    "value": "将进酒",
  },
  Object {
    "type": "text",
    "value": "君不见黄河之水天上来，奔流到海不复回。",
  },
  Object {
    "type": "text",
    "value": "
君不见高堂明镜悲白发，朝如青丝暮成雪。",
  },
  Object {
    "type": "text",
    "value": "
人生得意须尽欢，莫使金樽空对月。",
  },
  Object {
    "type": "text",
    "value": "
天生我材必有用，千金散尽还复来。",
  },
  Object {
    "type": "text",
    "value": "
烹羊宰牛且为乐，会须一饮三百杯。",
  },
  Object {
    "type": "text",
    "value": "
岑夫子，丹丘生，将进酒，杯莫停。",
  },
  Object {
    "type": "text",
    "value": "
与君歌一曲，请君为我倾耳听。",
  },
  Object {
    "type": "text",
    "value": "
钟鼓馔玉不足贵，但愿长醉不复醒。",
  },
  Object {
    "type": "text",
    "value": "
古来圣贤皆寂寞，惟有饮者留其名。",
  },
  Object {
    "type": "text",
    "value": "
陈王昔时宴平乐，斗酒十千恣欢谑。",
  },
  Object {
    "type": "text",
    "value": "
主人何为言少钱，径须沽取对君酌。",
  },
  Object {
    "type": "text",
    "value": "
五花马、千金裘，呼儿将出换美酒，与尔同销万古愁。",
  },
  Object {
    "type": "text",
    "value": "emphasis: ",
  },
  Object {
    "type": "text",
    "value": "foo \\"",
  },
  Object {
    "type": "text",
    "value": "bar",
  },
  Object {
    "type": "text",
    "value": "\\" foo",
  },
  Object {
    "type": "text",
    "value": "autolink(extension): ",
  },
  Object {
    "type": "text",
    "value": "http://example.com",
  },
  Object {
    "type": "text",
    "value": "autolink: ",
  },
  Object {
    "type": "text",
    "value": "foo@bar.example.com",
  },
  Object {
    "type": "text",
    "value": "Setext headings",
  },
  Object {
    "type": "text",
    "value": "Task List Items (extension):",
  },
  Object {
    "type": "text",
    "value": "foo",
  },
  Object {
    "type": "text",
    "value": "bar",
  },
  Object {
    "type": "text",
    "value": "baz",
  },
  Object {
    "type": "text",
    "value": "bim",
  },
  Object {
    "type": "text",
    "value": "This is a blockquote",
  },
  Object {
    "type": "text",
    "value": "with multiple lines.",
  },
  Object {
    "type": "text",
    "value": "This is a math block:",
  },
  Object {
    "type": "text",
    "value": "heading1",
  },
  Object {
    "type": "text",
    "value": "heading2",
  },
  Object {
    "type": "text",
    "value": "heading3",
  },
  Object {
    "type": "text",
    "value": "heading4",
  },
  Object {
    "type": "text",
    "value": "heading5",
  },
  Object {
    "type": "text",
    "value": "heading6",
  },
  Object {
    "type": "text",
    "value": "This is a footnote ",
  },
  Object {
    "type": "text",
    "value": "something....",
  },
  Object {
    "type": "text",
    "value": ".",
  },
  Object {
    "type": "text",
    "value": "This is a footnote definition.",
  },
  Object {
    "type": "text",
    "value": "This is a footnote reference ",
  },
  Object {
    "type": "text",
    "value": "Definitions:",
  },
  Object {
    "type": "text",
    "value": "logo",
  },
  Object {
    "type": "text",
    "value": "baz",
  },
  Object {
    "type": "text",
    "value": "
",
  },
  Object {
    "type": "text",
    "value": "logo",
  },
  Object {
    "type": "text",
    "value": " ",
  },
  Object {
    "type": "text",
    "value": "bar",
  },
  Object {
    "type": "text",
    "value": "
",
  },
  Object {
    "type": "text",
    "value": "logo",
  },
  Object {
    "type": "text",
    "value": "Guanghechen",
  },
  Object {
    "type": "text",
    "value": " is the author of yozora.",
  },
]
`;

exports[`collectNodes basic1 2`] = `
Array [
  Object {
    "children": Array [
      Object {
        "type": "text",
        "value": "君不见黄河之水天上来，奔流到海不复回。",
      },
      Object {
        "type": "break",
      },
      Object {
        "type": "text",
        "value": "
君不见高堂明镜悲白发，朝如青丝暮成雪。",
      },
      Object {
        "type": "break",
      },
      Object {
        "type": "text",
        "value": "
人生得意须尽欢，莫使金樽空对月。",
      },
      Object {
        "type": "break",
      },
      Object {
        "type": "text",
        "value": "
天生我材必有用，千金散尽还复来。",
      },
      Object {
        "type": "break",
      },
      Object {
        "type": "text",
        "value": "
烹羊宰牛且为乐，会须一饮三百杯。",
      },
      Object {
        "type": "break",
      },
      Object {
        "type": "text",
        "value": "
岑夫子，丹丘生，将进酒，杯莫停。",
      },
      Object {
        "type": "break",
      },
      Object {
        "type": "text",
        "value": "
与君歌一曲，请君为我倾耳听。",
      },
      Object {
        "type": "break",
      },
      Object {
        "type": "text",
        "value": "
钟鼓馔玉不足贵，但愿长醉不复醒。",
      },
      Object {
        "type": "break",
      },
      Object {
        "type": "text",
        "value": "
古来圣贤皆寂寞，惟有饮者留其名。",
      },
      Object {
        "type": "break",
      },
      Object {
        "type": "text",
        "value": "
陈王昔时宴平乐，斗酒十千恣欢谑。",
      },
      Object {
        "type": "break",
      },
      Object {
        "type": "text",
        "value": "
主人何为言少钱，径须沽取对君酌。",
      },
      Object {
        "type": "break",
      },
      Object {
        "type": "text",
        "value": "
五花马、千金裘，呼儿将出换美酒，与尔同销万古愁。",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "alt": "logo",
        "identifier": "logo",
        "label": "logo",
        "referenceType": "collapsed",
        "type": "imageReference",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "type": "text",
        "value": "emphasis: ",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "foo \\"",
          },
          Object {
            "children": Array [
              Object {
                "type": "text",
                "value": "bar",
              },
            ],
            "type": "emphasis",
          },
          Object {
            "type": "text",
            "value": "\\" foo",
          },
        ],
        "type": "strong",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "type": "text",
        "value": "autolink(extension): ",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "http://example.com",
          },
        ],
        "type": "link",
        "url": "http://example.com",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "type": "text",
        "value": "autolink: ",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "foo@bar.example.com",
          },
        ],
        "type": "link",
        "url": "mailto:foo@bar.example.com",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "type": "text",
        "value": "Task List Items (extension):",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "bar",
          },
        ],
        "status": "doing",
        "type": "listItem",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "baz",
          },
        ],
        "status": "done",
        "type": "listItem",
      },
    ],
    "marker": 45,
    "ordered": false,
    "spread": false,
    "type": "list",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "foo",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "bar",
                  },
                ],
                "status": "doing",
                "type": "listItem",
              },
              Object {
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "baz",
                  },
                ],
                "status": "done",
                "type": "listItem",
              },
            ],
            "marker": 45,
            "ordered": false,
            "spread": false,
            "type": "list",
          },
        ],
        "status": "done",
        "type": "listItem",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "bim",
          },
        ],
        "status": "todo",
        "type": "listItem",
      },
    ],
    "marker": 45,
    "ordered": false,
    "spread": false,
    "type": "list",
  },
  Object {
    "children": Array [
      Object {
        "type": "text",
        "value": "This is a blockquote",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "type": "text",
        "value": "with multiple lines.",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "type": "text",
        "value": "This is a math block:",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "type": "text",
        "value": "This is a footnote ",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "something....",
          },
        ],
        "type": "footnote",
      },
      Object {
        "type": "text",
        "value": ".",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "type": "text",
        "value": "This is a footnote definition.",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "type": "text",
        "value": "This is a footnote reference ",
      },
      Object {
        "identifier": "foo-2",
        "label": "foo-2",
        "type": "footnoteReference",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "type": "text",
        "value": "Definitions:",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "logo",
          },
        ],
        "identifier": "bar",
        "label": "bar",
        "referenceType": "full",
        "type": "linkReference",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "baz",
          },
        ],
        "identifier": "baz",
        "label": "baz",
        "referenceType": "shortcut",
        "type": "linkReference",
      },
      Object {
        "type": "text",
        "value": "
",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "logo",
          },
        ],
        "identifier": "logo",
        "label": "logo",
        "referenceType": "shortcut",
        "type": "linkReference",
      },
      Object {
        "type": "text",
        "value": " ",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "bar",
          },
        ],
        "identifier": "baz",
        "label": "baz",
        "referenceType": "full",
        "type": "linkReference",
      },
      Object {
        "type": "text",
        "value": "
",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "logo",
          },
        ],
        "identifier": "logo",
        "label": "logo",
        "referenceType": "collapsed",
        "type": "linkReference",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "alt": "logo",
        "identifier": "logo",
        "label": "logo",
        "referenceType": "collapsed",
        "type": "imageReference",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "Guanghechen",
          },
        ],
        "title": "this is a link",
        "type": "link",
        "url": "https://github.com/guanghechen",
      },
      Object {
        "type": "text",
        "value": " is the author of yozora.",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "alt": "foo",
        "title": "guanghechen",
        "type": "image",
        "url": "https://avatars.githubusercontent.com/u/42513619?v=4",
      },
    ],
    "type": "paragraph",
  },
]
`;
